app/services/fixity_checker.rb

```diff
@@ -0,0 +1,117 @@
+    raise ArgumentError.new("Please pass in an Asset.") unless asset.is_a? Asset
+    @asset = asset
+  end
+
+  #FixityChecker.new(asset).check
+  def check(sieve_integer = 0)
+    return unless sift(sieve_integer)
+    return nil if @asset.nil?
+    return nil if @asset.file.nil?
+    the_expected_checksum = expected_checksum
+    # Note: @asset.file.sha512 can be nil on e.g.
+    # items imported with DISABLE_BYTESTREAM_IMPORT=true
+    # For now, we are ignoring these.
+    return nil if the_expected_checksum.nil?
+
+    new_check = FixityCheck.new(asset: @asset)
+    new_check.expected_result = the_expected_checksum
+    new_check.checked_uri = @asset.file.url
+    new_check.actual_result = actual_checksum
+    new_check.passed= (the_expected_checksum==actual_checksum)
+    new_check.save!
+  end
+
+  #FixityChecker.new(asset).prune_checks
+  # After pruning, for each asset - URI combination, you will be left with:
+  # up to NUMBER_OF_RECENT_PASSED_CHECKS_TO_KEEP recent passed checks,
+  # plus all the failed checks,
+  # plus one initial check (to remember the file that was intially uploaded.)
+  def prune_checks(sieve_integer = 0)
+    return unless sift(sieve_integer)
+    return if @asset.file.nil?
+    return if @asset.file.url.nil?
+    checks_its_ok_to_delete.map(&:destroy)
+  end
+
+  private
+
+  # Returns an array of checks that we don't want to keep.
+  # The current rules are:
+  # Never throw out FAILED checks
+  # Never throw out the earliest PASSED check
+  # Always keep N recent PASSED checks
+  def checks_its_ok_to_delete
+    # Throw all the passed checks INTO the trash.
+    checks = FixityCheck.checks_for(@asset, @asset.file.url)
+    return [] if checks.empty?
+    trash = checks.select { | ch | ch.passed? }
+    # But then, pop the earliest passed check back OUT of the trash.
+    earliest_passed_check = trash.pop
+    # Finally, shift the most recent N passed checks back OUT of the trash.
+    recent_passed_checks = trash.shift(NUMBER_OF_RECENT_PASSED_CHECKS_TO_KEEP)
+    return trash
+  end
+
+  def expected_checksum
+    @asset.file.sha512
+  end
+
+
+  # Sifts all our  assets by some integer
+  # between one and CHECK_CYCLE_LENGTH.
+  # Allows us to convieniently check only a subset of
+  # the assets at a time, but be sure everything eventually
+  # gets checked.
+  # Pass in 0, and everything will go through the sieve.
+  def sift(sieve_integer)
+    return true if sieve_integer == 0
+    check_sieve(sieve_integer)
+    @asset.friendlier_id.bytes.sum % CHECK_CYCLE_LENGTH == sieve_integer
+  end
+
+  # Check this is a positive int between 1 and the cycle length.
+  def check_sieve(sieve_integer)
+    if !(sieve_integer.is_a? Integer)        ||
+      sieve_integer < 1                      ||
+      sieve_integer > CHECK_CYCLE_LENGTH
+      raise ArgumentError.new(
+        "Expected a positive int between  1 and #{CHECK_CYCLE_LENGTH}. Got #{sieve_integer}"
+      )
+    end
+  end
+
+  def actual_checksum
```

Comment:
This calculates a checksum every time you call it, it doens't have memoization. 

So if you call it twice, you'll calculate it twice. And for instance, you do call it twice when you create the FixityCheck record, once for setting as the value of `actual_result`, and once for checking `(expected_checksum == the_actual_checksum)` -- so you're calculating checksum twice for every check, unnecessarily! And it is a fairly expensive operation, the most expensive thing `check` is doing. 

So either memoize it in the `actual_checksum` method (which I think would be fine), or note in comments that it gets calculated on every call, _and_ adjust the `check` method (and anywhere else) to make sure to put the result in a local var and re-use it, instead of calculating twice. I think memoizing is fine, but either would work.

Reply:
Memoized!